apps/system/utils/utils_heapinfo.c

```diff
@@ -218,9 +221,19 @@ static void heapinfo_show_taskinfo(struct mm_heap_s *heap)
 #endif
 	printf("-------|-----------|-----------|----------\n");
 
+#if CONFIG_MM_NHEAPS > 1
+	int read_flag[CONFIG_MAX_TASKS] = {0};
+#endif
+
 	for (heap_idx = 0; heap_idx < CONFIG_MM_NHEAPS; heap_idx++) {
 		for (tcb_idx = 0; tcb_idx < CONFIG_MAX_TASKS; tcb_idx++) {
 			if (heap[heap_idx].alloc_list[tcb_idx].pid != HEAPINFO_INIT_INFO) {
+#if CONFIG_MM_NHEAPS > 1
+				if (read_flag[tcb_idx] == 1) {
```

Comment:
From the proc, current and peak heap are bounded with tcb per each heap.
So I think, it is not duplicated, it is different data.

Reply:
1. From the code, `heapinfo_read_proc(pid)` may be called several times in different heaps, and for same TCB, the `pid` is same, so the proc/*/stat output is same certainly.
This is a sample output: (**I config 6 regions in 4 heaps**)
```log
PID |  PPID | STACK | CURR_HEAP | PEAK_HEAP | NAME
----|-------|-------|-----------|-----------|----------
  0 |     0 |  4096 |     53008 |     53328 | Idle Task()
  1 |     0 |  4156 |         0 |        48 | hpwork()
  2 |     0 |  4156 |         0 |         0 | lpwork()
  3 |     0 |  4152 |         0 |      1664 | LWIP_TCP/IP()
  5 |     4 |  4156 |        96 |        96 | tash()
  8 |     7 |  4152 |      8288 |      8672 | eventTask()
  9 |     7 |  3644 |     16512 |     16848 | wifi()
 10 |    -1 |  3148 |         0 |         0 | IoT_MessageHandlerQueue()
 11 |    -1 |  2124 |         0 |         0 | IoT_Retransmit()
 12 |    -1 |  8268 |        48 |        48 | <pthread>()
 13 |    -1 |  3148 |         0 |         0 | IoT_IPSendQueue()
 14 |    -1 |  8268 |       144 |       144 | IoT_ReceiveHandler()
 15 |    -1 |  3148 |     38656 |     40576 | IoT_TCPSendQueue()
 16 |    -1 |  8268 |      6480 |     14304 | IoT_TCPReceive()
 17 |    -1 |  8268 |     12960 |     13776 | THINGS_STACK_SERVEREXCE_LOOP()
 18 |    -1 |  4172 |        48 |        48 | THINGS_STACK_CLOUD_TOKEN_CHECK()
 19 |    -1 |  8268 |         0 |        48 | button_scan_thread()
 21 |    20 |  2104 |        48 |        96 | NTP daemon()
 24 |    -1 |  4172 |        48 |        48 | THINGS_STACK_PING()
  0 |     0 |  4096 |     53008 |     53328 | Idle Task()
  1 |     0 |  4156 |         0 |        48 | hpwork()
  3 |     0 |  4152 |         0 |      1664 | LWIP_TCP/IP()
  5 |     4 |  4156 |        96 |        96 | tash()
Failed to open /proc/7/stat
Failed to read /proc/7/stat
  8 |     7 |  4152 |      8288 |      8672 | eventTask()
  9 |     7 |  3644 |     16512 |     16848 | wifi()
 10 |    -1 |  3148 |         0 |         0 | IoT_MessageHandlerQueue()
 13 |    -1 |  3148 |         0 |         0 | IoT_IPSendQueue()
 14 |    -1 |  8268 |       144 |       144 | IoT_ReceiveHandler()
 15 |    -1 |  3148 |     38656 |     40576 | IoT_TCPSendQueue()
 16 |    -1 |  8268 |      6480 |     14304 | IoT_TCPReceive()
 17 |    -1 |  8268 |     12960 |     13776 | THINGS_STACK_SERVEREXCE_LOOP()
Failed to open /proc/20/stat
Failed to read /proc/20/stat
 21 |    20 |  2104 |        48 |        96 | NTP daemon()
Failed to open /proc/22/stat
Failed to read /proc/22/stat
Failed to open /proc/7/stat
Failed to read /proc/7/stat
 17 |    -1 |  8268 |     12960 |     13776 | THINGS_STACK_SERVEREXCE_LOOP()
Failed to open /proc/20/stat
Failed to read /proc/20/stat
Failed to open /proc/22/stat
Failed to read /proc/22/stat
 24 |    -1 |  4172 |        48 |        48 | THINGS_STACK_PING()
Failed to open /proc/7/stat
Failed to read /proc/7/stat
 10 |    -1 |  3148 |         0 |         0 | IoT_MessageHandlerQueue()
 14 |    -1 |  8268 |       144 |       144 | IoT_ReceiveHandler()
 15 |    -1 |  3148 |     38656 |     40576 | IoT_TCPSendQueue()
 16 |    -1 |  8268 |      6480 |     14304 | IoT_TCPReceive()
 17 |    -1 |  8268 |     12960 |     13776 | THINGS_STACK_SERVEREXCE_LOOP()
Failed to open /proc/22/stat
Failed to read /proc/22/stat
 24 |    -1 |  4172 |        48 |        48 | THINGS_STACK_PING()
```
e.g. Pid 17 stat is printed once for each heap (4 times in total), and they are same.

In other words, proc/*/stat should be only related to a certain task, no matter how many heaps it allocated memory from. isn't it?

Reply:
2. You said `From the proc, current and peak heap are bounded with tcb per each heap.`,
Did you check below code? right?
```c
ssize_t proc_entry_stat(...)
{
	...
	curr_heap = -1;
	peak_heap = -1;
#ifdef CONFIG_DEBUG_MM_HEAPINFO
	if (tcb->pid == 0) {
		/* Idle task normally uses heap with index 0. */
		heap = mm_get_heap_with_index(0);               ///////////// heap 0
	} else {
		heap = mm_get_heap(tcb->stack_alloc_ptr);    /////////////  also heap 0
	}
	if (heap == NULL) {
		return -1;
	}

	hash_pid = PIDHASH(tcb->pid);
	if (heap->alloc_list[hash_pid].pid == tcb->pid) {
		curr_heap = heap->alloc_list[hash_pid].curr_alloc_size;
		peak_heap = heap->alloc_list[hash_pid].peak_alloc_size;
	}
#endif
	...
}
```
I think there's another problem:
From above code, we can see that the current and peak heap are always got from heap 0 (currently we always allocate task stack from heap 0), but one task may also allocate memory from other heaps directly or indirectly, just like that Pid 17 task from all heaps. So I think this current and peak heap statistic is not exact, we should calculate (add up) heap in all heaps...
What's your opinion?

Reply:
Yes. You are right. I just saw the below..
``
if (heap->alloc_list[hash_pid].pid == tcb->pid) {
		curr_heap = heap->alloc_list[hash_pid].curr_alloc_size;
		peak_heap = heap->alloc_list[hash_pid].peak_alloc_size;
	}
```
We should modify the proc logic. 
I think, if we modify the proc, then there's no duplication. Right?

Reply:
No, duplication still exists, they are different problem.

Reply:
And I'm fixing duplication issue in this PR.
About proc logic, I will modify and test it...

Reply:
Why? Every heap has curr and peak size which bounded with itself per tcb. 
There can be duplicated pid in the heapinfo output list, but the contents will be different.
Uhm... yeah..... we should also change how to show the summation of heap usage per tcb, if a task/pthread uses multiple heap...

Reply:
Pid may be duplicated in different heaps, but I don't understand why the contents will be different?
You can see above sample output, 4 records are completely same:
17 |    -1 |  8268 |     12960 |     13776 | THINGS_STACK_SERVEREXCE_LOOP()
17 |    -1 |  8268 |     12960 |     13776 | THINGS_STACK_SERVEREXCE_LOOP()
17 |    -1 |  8268 |     12960 |     13776 | THINGS_STACK_SERVEREXCE_LOOP()
17 |    -1 |  8268 |     12960 |     13776 | THINGS_STACK_SERVEREXCE_LOOP()
Because  `heapinfo_read_proc(17)` was called 4 times (once per heap)

Reply:
My summary:
1. There can be duplicated pid in multiple heaps, because each heap maintain it's own `heapinfo_tcb_info_t alloc_list[CONFIG_MAX_TASKS]`.
2. heapinfo_read_proc(pid) should only be called once for each pid, even if the pid is duplicated in different heaps. (this PR #3088)
3. heapinfo_read_proc() should add up heap usage in all heaps for per pid/tcb. (#3090)

@jeongchanKim Please help to review #3090